### DISCUSSION TITLE
OAuth / Onboard User with additional Meta-Data

### DISCUSSION THREAD

**Discussion by aheissenberger** [zitadel/zitadel#7775] [Q&A]
Is there an Option to add Meta-Data to a User registered by the default Zitadel Registration Form as part of the call to initialize the registration process with `prompt=create`.

Or can I add a custom scope e.g. `urn:custom:meta:<key>:<value>` which is read by an action after registration and converted to a meta value on the user.

**Comment by fforootd** [comment]
An action can access the scope you sent along https://zitadel.com/docs/apis/actions/objects#auth-request

So you should be able to pull that scope in and store metadata accordingly.

**Reply by aheissenberger** [reply]
I wrote an action to log all values I get in by the pre-register trigger but I was not able to add variable content. My custom scope e.g. "urn:custom:meta:<key>:<value>" was removed from the data which was provided to the action. The doc for scopes say that custom scopes are planned for a future release.

We also tried to solve this with the state parameter in the OAuth process but this does not work if the user is using the link in the email verification email https://github.com/zitadel/zitadel/discussions/7779#discussioncomment-9139083

**Reply by mcblair** [reply]
@fforootd Are you suggesting that currently I can provide custom scopes via `request.oidc.scopes[]` and use that scope with a pre-access token action? This would be perfect for us as a B2B platform.

My current use case is to use an action to call out to our Claims Value Provider (CVP) to return domain-derived user claims. Consider the two scenarios:
A: User with role 'Business Admin' calls an API `GET /things` 
1. Authorize occurs, pre-access token action runs, calls our CVP
2. Our CVP returns `businessId` claim and sets it on access token
3. Our `things` API authorizes the request based on the `businessId` claim, and then returns `things` data filtered by that `businessId`.

B: User with role 'Business Admin' calls an API `GET /things/{thingId}`
1. Authorize occurs, pre-access token action runs, calls our CVP
2. Our CVP returns `thingId` cliam and sets it on access token
3. Our `things` API authorizes the request based on the `thingId` claim, and then returns the `thing` data. 

In the first case and other coarse-grained cases, the approach works as-is - but our challenge right now is in the finer-grained cases like the second case. If Zitadel can support this, Zitadel will be our IDP choice.

**Reply by fforootd** [reply]
@mcblair of the top of my head I would say that should work but do not hold me too it 😁 

But I think you need two steps ATM (actions v2 will change this to be more easy)

1) Store the scope as metedata
2) Consume the metadata in the token creation step

**Reply by dmlang** [reply]
@aheissenberger have you been able to solve the problem with the custom scope accessibility in an action?

**Reply by aheissenberger** [reply]
> @aheissenberger have you been able to solve the problem with the custom scope accessibility in an action?

@dmlang no

**Reply by mcblair** [reply]
> @mcblair of the top of my head I would say that should work but do not hold me too it 😁
> 
> But I think you need two steps ATM (actions v2 will change this to be more easy)
> 
>     1. Store the scope as metedata
> 
>     2. Consume the metadata in the token creation step

We have a use case in which a user does not exist yet, but APIs need to be secured (a user is applying for an account). We first use Twilio Verify for OTP, and as a result of a successful OTP we'd like our backend service to issue an access token and return it for our UI to use when submitting the application. The access token would ideally be requested with a narrow scope and be issued with a custom claim so that its access is limited. The backend service that we want to issue the access token currently uses a service user JWT to create a Zitadel user as a result of a successful application. 

Is there a recommended approach for this? My understanding right now is that our service user authenticated backend service should be able to use the Token Exchange feature to achieve this but would still need a Zitadel Action (i'm not sure which flow and trigger it would belong?) to handle the custom scope and set the appropriate claims. @fforootd what do you think?
